### PR TITLE
[WIP] [CR] Robot-hauler companion

### PR DESCRIPTION
**Pull request by Xelat84**
#### Summary

```SUMMARY: Features "Robot-hauler companion which helps transporting items, especially between levels with z-levels option enabled"```

#### Purpose of change

Main purpose is make transporting items from underground lab/mansion levels less tedious.

#### Describe the solution

Player will be able to craft a robot and attach some special container to it so this robot will be able to hold a lot of weight and capacity. It will act like a tamed dog with attached bag.

### CODE REVIEW

**Comment by PenceGL** [comment]
This is a cool idea. Are you going to make it tough to obtain? It seems like it should definitely be a late game type of tech with some high electronics/fabrication requirements to create.

It seems like it could open up some really cool possibilities down the line too... if it was treated like a vehicle it could be modified to have weapons mounted onto it. Or there could just be variants, like a storage bot, a defender bot that's got a specific weapon, crafting station bot that follows you around carrying welding or even cooking appliances, etc.

**Comment by cainiaowu** [comment]
For transporting between Z level, why cant wen make a makeshift rope and pulley system that work as a hoist, build on site.
That technology have already been existed since Pharaoh times.

**Comment by Xelat84** [comment]
> For transporting between Z level, why cant wen make a makeshift rope and pulley system that work as a hoist, build on site.
That technology have already been existed since Pharaoh times.

Yes, I already mentioned that in another discussion about a year ago or so, and have plans to implement it too, but this one looks for me little bit simplier and faster at implementation.

> I'd put this at 10/10/10 electronics/computers/fabrication, and for most it would be salvageable instead of constructable.

Sure it's discussable.

**Comment by CoroNaut** [comment]
So its basically a shopping cart that follows you automatically and climbs stairs? Neat. Will it be strictly an entity or allow modification similar to a vehicle? Perhaps once we get a decent use for the Robot driving unit, we can make vehicles that follow but not go up stairs?

**Comment by Regilith** [comment]
It might be a better bet to have the player repurpose or repair an existing drone for this, such as the riot police bot / eyebot corpse items and costing 7-8 computers & electronics, versus building one from scratch and requiring an extremely late game 10 10 10. You could also do both, have the repurposed one just have worse speed or carrying capacity versus the custom built one.

**Comment by Xelat84** [comment]
Actually found an image of almost what I imagine just before started working on this.
And honestly I believe there is no reason to make so high skill requirements to craft such tech. It's really pretty simple!
Btw I also initially planned to make it able to hold 1000kg of weight. What a coincidence? :)
carrier_1000kg

**Comment by PenceGL** [comment]
Why do you think it shouldn't have a high requirement? There's been plenty of talk in other threads about stuff like CBMs, for example. We have had CBM crafting removed because it's delicate, super high-tech electronics that would not be feasible for a DIY kind of craft.

The robot in your picture is not much different. An on top of that it's not a passive device, it's going to have AI that allows it to track the user and follow them. Definitely not easy to craft from scrap materials.

**Comment by Shodan14** [comment]
[There's also stuff like this](https://mms.businesswire.com/media/20160525005864/en/526869/5/THeMIS_during_Spring_Storm.jpg)

**Comment by Xelat84** [comment]
Finished scripts and configs.
Now we can discuss recipes and skill requirements. I believe I setup everything properly though.

Also will appreciate for suggestions in which drop lists add robot parts. For now I added some to regular "robots" drop list, and some parts in specific bots drop lists.

File: data/json/items/generic.json [hunk 1]
```diff
@@ -2496,5 +2496,82 @@
     "symbol": ".",
     "color": "white",
     "ammo_type": "components"
+  },
+  {
+    "id": "hp_compressor",
+    "copy-from": "pump_complex",
```
Comment (Night-Pryanik) [inline data/json/items/generic.json]: Why do we need another high pressure pump with slightly different characteristics? To be able to craft it, as existing high pressure pump isn't craftable on purpose?

File: data/json/items/generic.json [hunk 2]
```diff
@@ -2496,5 +2496,82 @@
...
+    "price": 47900
+  },
+  {
+    "type": "GENERIC",
+    "id": "hydraulic_platform",
+    "symbol": ";",
+    "color": "light_gray",
+    "name": "hydraulically adjustable platform",
+    "category": "other",
+    "description": "Steel frames with two hydraulic jacks, capable of lifting and inclining up to 1000kg using electrical power.",
```
Comment (Night-Pryanik) [inline data/json/items/generic.json]: 1000kg -> 1000 kg.

File: data/json/recipes/recipe_robots.json [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+      [ [ "hp_compressor", 1 ] ],
+      [ [ "hose", 2 ] ],
+      [ [ "spring", 4 ] ],
+      [ [ "plastic_chunk", 4 ] ],
+      [ [ "steel_lump", 1 ], [ "steel_chunk", 4 ], [ "scrap", 16 ] ]
+    ]
+  },
+  {
+    "type": "recipe",
+    "result": "bot_navigation",
```
Comment (Night-Pryanik) [inline data/json/recipes/recipe_robots.json]: Craftable full AI? I won't believe it.

Comment (Xelat84) [reply data/json/recipes/recipe_robots.json]: Sorry, it's not an AI at all, I'll give it more correct name. It's just simple guidance unit, only to be enough to follow player character.

File: data/json/items/generic.json (hunk 1 above)
Comment (Xelat84) [reply data/json/items/generic.json]: Definitely no. This one is electric compressor, that uses existing (mechanic) one, which is not craftable as you notice, in it's own recipe, if you look closely.

**Comment by Xelat84** [comment]
>Minireactor on a robot while most vehicles still use gas, diesel or solar energy?

Not sure what's wrong with it. You can obtain plutonium cells from killing existing robots, it's intended that they use reactor for getting energy from plutonium cells, am I wrong?

Also it's mostly 'temporary' solution, as implementing consuming fuel for this robot is a little bit complex for this iteration. So I decide to use minireactor and plutonium cells as most reliable solution, that can work long enough without asking why it's not consuming any fuel continously.
May be later I will add version with gas/diesel engine. May be even with solar panels. For now I want to leave robot power system as is.

**Comment by mrkybe** [comment]
Will it be able to ride in a seat in a death mobile, or does it need an carrier thingy?

**Comment by Xelat84** [comment]
> Will it be able to ride in a seat in a death mobile, or does it need an carrier thingy?

I'm not sure about seat, I didn't expect that. But it should fit in a large carrier, and also you can switch it off, so it folds and become just an item.

**Comment by kevingranade** [comment]
> Actually found an image of almost what I imagine just before started working on this.
And honestly I believe there is no reason to make so high skill requirements to craft such tech. It's really pretty simple!

It is simple, but it is *not what you implemented*. If you're making a powered load hauler like that, it looks like your properties are going to be:
Only navigates on flat ground or stairs
Moves under the direct control of the player (probably via a grab).
Still fairly difficult to craft (fabrication 5 or so).
Uses conventional power, either gasoline generator or batteries.

In short it's a powered shopping cart that can climb stairs, which IMO is a great fit for what you've said you're trying to accomplish.

What you've *actually* implemented is a perpetual motion all-terrain autonomous robot, which is something we spawn in a lab, but would definitely not be player craftable.

**Comment by kevingranade** [comment]
> Actually found an image of almost what I imagine just before started working on this.
And honestly I believe there is no reason to make so high skill requirements to craft such tech. It's really pretty simple!

It is simple, but it is *not what you implemented*. If you're making a powered load hauler like that, it looks like your properties are going to be:
Only navigates on flat ground or stairs
Moves under the direct control of the player (probably via a grab).
Still fairly difficult to craft (fabrication 5 or so).
Uses conventional power, either gasoline generator or batteries.

In short it's a powered shopping cart that can climb stairs, which IMO is a great fit for what you've said you're trying to accomplish.

What you've *actually* implemented is a perpetual motion all-terrain autonomous robot, which is something we spawn in a lab, but would definitely not be player craftable.

**Comment by SunshineDistillery** [comment]
I'll share how I handled it in a robot mod I'm working on :
- Define a corpse item called broken_robot_hauler. 
- Put the corpse in item groups and/or spawn them as hostile mon_robot_hauler.
- Make a recipe for bot_robot_hauler that has the corpse item as the first component. 

That way the player isn't building a super complex robot from scratch but is just fixing up an existing bot. Basically repairing an item through the crafting menu.  It also means it requires a rare item AND a minimum skill level, thus pushing robot pals back to mid or late game (or a start dedicated to repair skills).  I also added an electrohack to the tools to imply that there is some reprogramming or hot-wiring going on.

**Comment by Xelat84** [comment]
>Define a corpse item called broken_robot_hauler.
Put the corpse in item groups and/or spawn them as hostile mon_robot_hauler.
Make a recipe for bot_robot_hauler that has the corpse item as the first component.

It's all already done, except adding broken version of whole robot to item groups for spawn. Look at commits. :)

But I agree with Kevin's point and perhaps I will try to make it walk only flat surface and work on consumable fuel/energy.

And in general I want to make all it parts craftable (but not from scratch), and only after that I will try using existing entities, like eyebot etc. and their parts as a side way to achieve this robot hauler.
